url2bc31e3489_ramonli_RabbitMQResearch_tgz-pJ8-com_mpos_lottery_te_thirdpartyservice_amqp_protobuf_CancellationMessagePulisherJ8/src/com/mpos/lottery/te/thirdpartyservice/amqp/RabbitMessagePublishMain.java

```diff
@@ -0,0 +1,245 @@
+             * <p>
+             * basic.nack will only be delivered if an internal error occurs in
+             * the Erlang process responsible for a queue.
+             * <p>
+             * refer to http://www.rabbitmq.com/confirms.html
+             */
+            @Override
+            public void handleNack(long deliveryTag, boolean multiple) throws IOException {
+                logger.info("Nack: " + deliveryTag);
+                // RabbitMessagePublishMain.this.release(connection);
+            }
+
+        });
+
+        /**
+         * If set 'immediate' to true, the publisher will throw
+         * " com.rabbitmq.client.AlreadyClosedException: clean connection shutdown; reason: Attempt to use closed channel"
+         * , as the server will immediately close the connection once received
+         * message.
+         * <p>
+         * Diff it with the case where no 'immediate' set, in fact RabbitMQ
+         * doesn't support this parameter.
+         */
+        logger.debug("Prepare to publish message.");
+        channel.basicPublish(exchangeName, routingKey, props, message);
+        logger.debug("Publish message successfully");
+
+        /**
+         * Why my returnListener is not invoked?
+         * <p>
+         * http://rabbitmq.1065348.n5.nabble.com/ReturnListener-is-not-invoked-
+         * td24549.html
+         * <p>
+         * The call to Channel#waitForConfirmsOrDie() will wait until all
+         * messages published since the last call have been either ack'd or
+         * nack'd by the broker.
+         * <p>
+         * Must call channel.confirmSelecte() first to enable confirm mode on
+         * channel.
+         */
+        logger.debug("Wait for confirmation message");
+        channel.waitForConfirmsOrDie(30*1000);
+        logger.debug("Message confirmed");
+        // now we can close connection
+        connection.close();
+    }
+
+    private void release(Connection conn) {
+        try {
+            if (conn != null)
+                conn.close();
+        }
+        catch (Exception e) {
+            logger.warn(e);
+        }
+    }
+
+    public String getHost() {
+        return host;
+    }
+
+    public static void main(String[] argv) throws Exception {
+        RabbitMessagePublishMain main = new RabbitMessagePublishMain("192.168.2.158", "/te", 5672, "amqp",
+                "amqp");
+        main.publish(EXCHANGE_NAME, EXCHANGE_NAME + ".-1", MessageProperties.PERSISTENT_BASIC,
+                "hello, ramon".getBytes());
+        logger.debug(" [x] Published message successfully.");
+        // deleteExchange(EXCHANGE_NAME);
+    }
+
+    public static void deleteExchange(String exchangeName) throws Exception {
+        ConnectionFactory factory = new ConnectionFactory();
+        factory.setHost("192.168.100.68");
+        factory.setVirtualHost("/");
+        factory.setUsername("amqp");
+        factory.setPassword("amqp");
+        Connection connection = factory.newConnection();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *connection*. The resource is referenced at line 240. The resource is closed at line 242. There are other execution paths that do not contain closure statements, for example, when Connection.createChannel throws an exception. Ensure *connection* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).